File: 859-buddy-strings.py
class Solution:
    def buddyStrings(self, s: str, goal: str) -> bool:
        if len(s) != len(goal):
            return False
        if s == goal:
            return len(set(s)) < len(s)

        dif = []
        for i in range(len(s)):
            if s[i] != goal[i]:
                dif.append(i)
        
        if len(dif) != 2:
            return False
        
        i, j = dif
        return s[i] == goal[j] and goal[i] == s[j]

File: 859-buddy-strings.java
class Solution {
    public boolean buddyStrings(String s, String goal) {

        if (s.length() != goal.length())
            return false;

        if (s.equals(goal)) {
            Set<Character> set = new HashSet<>();

            for (char ch : s.toCharArray()) {
                if (!set.add(ch))
                    return true;
            }

            return false;
        }

        int dif = 0;
        int first = -1;
        int second = -1;

        for (int i = 0; i < s.length(); i++) {

            if (s.charAt(i) != goal.charAt(i)) {
                dif++;

                if (first == -1)
                    first = i;
                else
                    second = i;
            }
        }

        if (dif != 2)
            return false;

        return s.charAt(first) == goal.charAt(second)
            && s.charAt(second) == goal.charAt(first);
    }
}
